File: src/K53Guru/src/Domain/Enums/AttemptMode.cs
namespace K53Guru.Domain.Enums;

/// <summary>
/// Which of the two learner experiences an <see cref="Entities.Attempt"/> is sitting under,
/// chosen once at <c>StartAttemptCommand</c> time and never changed thereafter (Story 3.6).
/// Practice mode permits immediate per-question correctness/explanation (via
/// <c>CheckAnswerCommand</c>) and retry, with no server-enforced time limit. Test mode withholds
/// correctness/explanations entirely until submission and enforces the sitting's configured
/// time limit server-side.
/// </summary>
public enum AttemptMode
{
    Practice,
    Test
}

File: src/k53_guru_app/lib/domain/attempt_mode.dart
/// Mirrors the backend's `AttemptMode` domain enum
/// (`src/K53Guru/src/Domain/Enums/AttemptMode.cs`), serialized as a string
/// by the server's global `JsonStringEnumConverter`.
enum AttemptMode {
  practice,
  test;

  String toJson() {
    switch (this) {
      case AttemptMode.practice:
        return 'Practice';
      case AttemptMode.test:
        return 'Test';
    }
  }

  static AttemptMode fromJson(String value) {
    switch (value.trim()) {
      case 'Practice':
        return AttemptMode.practice;
      case 'Test':
        return AttemptMode.test;
      default:
        throw FormatException('Unknown AttemptMode: $value');
    }
  }
}
